File: src/cspfj/constraint/MatrixGeneral.java
package cspfj.constraint;

import java.math.BigInteger;
import java.util.Arrays;

public class MatrixGeneral implements Matrix, Cloneable {

	public int[] skip;

	public boolean[] matrix;

	private final static BigInteger MAX_VALUE = BigInteger
			.valueOf(Integer.MAX_VALUE);

	public MatrixGeneral(int[] sizes, boolean initialState) {
		int nbValues = 1;
		for (int size : sizes) {
			nbValues *= size;
		}

		matrix = new boolean[nbValues];
		Arrays.fill(matrix, initialState);

		skip = new int[sizes.length];
		skip[0] = 1;
		for (int i = 1; i < sizes.length; i++) {
			skip[i] = skip[i - 1] * sizes[i - 1];
		}

		// System.out.println(Arrays.toString(skip));
	}

	@Override
	public boolean check(final int[] tuple) {
		return matrix[matrixIndex(tuple)];
	}

	public static boolean usable(final int[] sizes) {
		BigInteger nbValues = BigInteger.ONE;

		for (int size : sizes) {
			nbValues = nbValues.multiply(BigInteger.valueOf(size));
			if (nbValues.compareTo(MAX_VALUE) > 0) {
				return false;
			}
		}
		return true;
	}

	@Override
	public void set(final int[] tuple, final boolean status) {
		matrix[matrixIndex(tuple)] = status;
	}

	private int matrixIndex(final int[] tuple) {
		final int[] skip = this.skip;
		int index = 0;
		for (int i = skip.length; --i >= 0;) {
			index += skip[i] * tuple[i];
		}
		return index;
	}

	public MatrixGeneral clone() throws CloneNotSupportedException {
		final MatrixGeneral matrix = (MatrixGeneral) super.clone();
		matrix.skip = skip.clone();
		matrix.matrix = this.matrix.clone();
		return matrix;

	}

}

File: src/main/scala/concrete/constraint/extension/MatrixGeneral.scala
package concrete.constraint.extension

;

import com.typesafe.scalalogging.LazyLogging

final class MatrixGeneral(sizes: Array[Int], offsets: Array[Int], initialState: Boolean) extends Matrix with LazyLogging {

  val size: Int = {
    val nbValues = sizes.foldLeft(BigInt(1))(_ * _)
    require(nbValues.isValidInt)
    nbValues.toInt
  }
  private val skip = new Array[Int](sizes.length - 1)
  private var matrix = {
    Array.fill(size)(initialState)
  }

  skip(0) = sizes(0)
  for (i <- 1 until skip.length) {
    skip(i) = skip(i - 1) * sizes(i)
  }
  private var cardinality: Int = if (initialState) size else 0

  override def check(tuple: Array[Int]): Boolean = matrix(matrixIndex(tuple))

  override def set(tuple: Array[Int], status: Boolean): Unit = {
    val index = matrixIndex(tuple)
    val oldStatus = matrix(index)
    if (oldStatus != status) {
      matrix(index)
      if (status) {
        cardinality += 1
      } else {
        cardinality -= 1
      }
    }
  }

  private def matrixIndex(tuple: Array[Int]): Int = {
    var index = tuple(0) - offsets(0)
    for (i <- 0 until skip.length) {
      index += skip(i) * (tuple(i + 1) - offsets(i + 1))
    }
    index
  }

  def copy = {
    val matrixG = new MatrixGeneral(sizes, offsets, initialState)
    matrixG.matrix = this.matrix.clone
    matrixG
  }

  override def toString =
    s"sizes: ${sizes.mkString("[", ", ", "]")}, offsets: ${offsets.mkString("[", ", ", "]")}"

  def isEmpty = cardinality == 0

  def allowed: Iterator[Array[Int]] = {
    ???
    matrix.indices.iterator.filter(matrix).map { i =>
      Array(i)
    }
  }

}
